File: c++/diameter-of-binary-tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxDiameter = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        if (root == nullptr) return 0;
        if (root->left == nullptr && root->right == nullptr) return 0;
        diameter(root);
        return maxDiameter;
    }
    int diameter(TreeNode* root) {
        if (root == nullptr) return 0;
        auto left = diameter(root->left);
        auto right = diameter(root->right);
        maxDiameter = max(maxDiameter, left + right);
        return 1 + max(left, right);
    }
};

File: python/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = 0

        def _diameter(root: Optional[TreeNode]) -> None:
            if not root:
                return 0

            left_path = _diameter(root.left)
            right_path = _diameter(root.right)

            nonlocal diameter
            diameter = max(diameter, left_path + right_path)

            return max(left_path, right_path) + 1

        _diameter(root)
        return diameter
